### PR TITLE
update object acl

### PR DESCRIPTION
**Pull request by clarisse-lau**
- Update bucket policy to enforce granting the bucket owner full control
- Remove Synapse arn from deny statement
- Grant Synapse read access in lambda

### CODE REVIEW

File: s3_synapse_sync/lambda_function.py [hunk 1]
```diff
@@ -50,6 +50,8 @@ def create_filehandle(syn, event, filename, bucket, key, project_id):
     md5 = get_md5(event, header, bucket, key)
     file_id = syn.findEntityId(filename, parent)
 
+    boto3.resource('s3').ObjectAcl(bucket, key).put(GrantRead='id=d9df08ac799f2859d42a588b415111314cf66d0ffd072195f33b921db966b440')
```
Comment (marcomarasca) [inline s3_synapse_sync/lambda_function.py:53]: We might want to move this call in the conditional block below (line 58). This seems to unconditionally perform a request for updating the ACL of the object, maybe we could make the call only if the file is not linked to synapse?

**Review by marcomarasca — CHANGES_REQUESTED** [review]
LGTM, just added one comment.

File: s3_synapse_sync/lambda_function.py (hunk 1 above)
Comment (clarisse-lau) [reply s3_synapse_sync/lambda_function.py]: Thanks @marcomarasca. Makes sense, moved to the block indicated

**Review by marcomarasca — APPROVED** [review]

File: s3-synapse-sync-bucket.j2 [hunk 2]
```diff
@@ -137,16 +137,12 @@ Resources:
...
-                  - list: !Join [",",!Ref "S3UserARNs"]
+                - !Ref S3UserARNs
             Action: s3:PutObject
             Resource: !Sub "${S3Bucket.Arn}/*"
             Condition:
-              StringNotEqualsIgnoreCase:
-                s3:x-amz-grant-full-control: !Sub "id=${SynapseCanonicalId},id=${S3CanonicalUserId}"
               StringNotEquals:
-                s3:x-amz-grant-full-control: !Sub "id=${SynapseCanonicalId}, id=${S3CanonicalUserId}"
+                s3:x-amz-acl: bucket-owner-full-control
```
Comment (zaro0508) [inline s3-synapse-sync-bucket.j2:145]: if we don't need `SynapseCanonicalId` and `S3CanonicalUserId` then we should remove it from the list of parameters.

File: s3_synapse_sync/lambda_function.py [hunk 3]
```diff
@@ -58,6 +58,8 @@ def create_filehandle(syn, event, filename, bucket, key, project_id):
         contentType = mimetypes.guess_type(filename, strict=False)[0]
         storage_id = syn.restGET("/projectSettings/"+project_id+"/type/upload")['locations'][0]
 
+        boto3.resource('s3').ObjectAcl(bucket, key).put(GrantRead='id=d9df08ac799f2859d42a588b415111314cf66d0ffd072195f33b921db966b440')
```
Comment (zaro0508) [inline s3_synapse_sync/lambda_function.py:61]: if you think there's a chance this magic number would change then you should pass it in as an environment variable

**Review by zaro0508 — CHANGES_REQUESTED** [review]

File: s3-synapse-sync-bucket.j2 [hunk 4]
```diff
@@ -137,16 +137,12 @@ Resources:
             Principal:
               AWS: !Split
                 - ","
-                - !Sub
-                  - "${list},${SynapseRootArn}"
-                  - list: !Join [",",!Ref "S3UserARNs"]
+                - !Ref S3UserARNs
```
[the remark this answers is not in the document]
Comment (clarisse-lau) [reply s3-synapse-sync-bucket.j2]: Thanks, fixed.

File: s3-synapse-sync-bucket.j2 (hunk 2 above)
Comment (clarisse-lau) [reply s3-synapse-sync-bucket.j2:131]: Removed

File: s3_synapse_sync/lambda_function.py (hunk 3 above)
Comment (clarisse-lau) [reply s3_synapse_sync/lambda_function.py]: Fixed

File: README.md [hunk 5]
```diff
@@ -63,6 +63,7 @@ The lambda requires the environment variable `BUCKET_VARIABLES`: a yaml-format s
 
 s3-synapse-sync-bucket-vars.yaml:
 ```yaml
+SynapseCanonicalId: d9df08ac799f2859d42a588b415111314cf66d0ffd072195f33b921db966b440
```
Comment (zaro0508) [inline README.md:66]: this doesn't seem like a bucket variable, instead it's a global variable specific to the lambda.  How about we pass in a separate `SYNAPSE_CANONICAL_ID` env var here https://github.com/Sage-Bionetworks/s3-synapse-sync/blob/master/template.yaml#L60 ?

Comment (zaro0508) [reply README.md:66]: also please update corresponding docs

File: s3-synapse-sync-bucket.j2 [hunk 6]
```diff
@@ -21,18 +21,6 @@ Parameters:
     Default: "arn:aws:iam::325565585839:root"
     ConstraintDescription: >-
       Must be the ARN of the Synapse root account
-  SynapseCanonicalId:
```
Comment (zaro0508) [inline s3-synapse-sync-bucket.j2:24]: please update corresponding docs

**Review by zaro0508 — CHANGES_REQUESTED** [review]

**Comment by zaro0508** [comment]
we cannot merge this change until we update the corresponding deployment side files (scicomp-provisioner repo) otherwise the deployment side will fail. Please push PRs with updates to these files as necessary..
* https://github.com/Sage-Bionetworks/scicomp-provisioner/blob/master/config/prod/s3-synapse-sync.yaml
* https://github.com/Sage-Bionetworks/scicomp-provisioner/blob/master/config/prod/htan-dcc-bucket-a.yaml
* https://github.com/Sage-Bionetworks/scicomp-provisioner/blob/master/config/prod/htan-dcc-center-a.yaml
* https://github.com/Sage-Bionetworks/scicomp-provisioner/blob/master/data/s3-synapse-sync-bucket-vars.yaml

File: README.md (hunk 5 above)
Comment (clarisse-lau) [reply README.md]: Changed to global env var

**Comment by clarisse-lau** [comment]
@zaro0508 update as in first remove the existing lambda and buckets?

> we cannot merge this change until we update the corresponding deployment side files (scicomp-provisioner repo) otherwise the deployment side will fail. Please push PRs with updates to these files as necessary..
> 
> * https://github.com/Sage-Bionetworks/scicomp-provisioner/blob/master/config/prod/s3-synapse-sync.yaml
> * https://github.com/Sage-Bionetworks/scicomp-provisioner/blob/master/config/prod/htan-dcc-bucket-a.yaml
> * https://github.com/Sage-Bionetworks/scicomp-provisioner/blob/master/config/prod/htan-dcc-center-a.yaml
> * https://github.com/Sage-Bionetworks/scicomp-provisioner/blob/master/data/s3-synapse-sync-bucket-vars.yaml

**Comment by zaro0508** [comment]
> @zaro0508 update as in first remove the existing lambda and buckets?

These changes do not require deleting and re-creating of buckets.  If you want to delete something, like `htan-dcc-bucket-a.yaml` then push a PR to delete.  If you want to keep it then push a PR with the needed updates.

**Review by zaro0508 — APPROVED** [review]

**Pull request merged**